# Librerías y configuraciones


In [1]:
import pandas as pd


In [2]:
pd.set_option('display.max_colwidth', 100)

---

# Carga de datos

In [3]:
df_genres = pd.read_csv('data/df_genres.csv')
df_item_genre = pd.read_csv('data/df_item_genre.csv')
df_items = pd.read_csv('data/df_items.csv')
df_reviews = pd.read_csv('data/df_reviews.csv')
df_users_items = pd.read_csv('data/df_users_items.csv')
df_users = pd.read_csv('data/df_users.csv')

----

Funciones auxiliares

In [4]:
def items_faltantes(columna_principal,columna_relacionada):
    list_elementos_unicos_de_columna_principal = list(columna_principal.unique())
    list_elementos_unicos_de_columna_relacionada = list(columna_relacionada.unique())
    print(len(list_elementos_unicos_de_columna_principal))
    print(len(list_elementos_unicos_de_columna_relacionada))

    # Convertir las listas a conjuntos
    list_elementos_unicos_de_columna_principal = set(list_elementos_unicos_de_columna_principal)
    list_elementos_unicos_de_columna_relacionada = set(list_elementos_unicos_de_columna_relacionada)

    # Encontrar los elementos que están en set_items_en_df_item_genre pero no en set_items_en_df_items
    elementos_no_en_df_items =   list_elementos_unicos_de_columna_relacionada - list_elementos_unicos_de_columna_principal
    # Obtener la cantidad y los elementos que no están en df_items
    cantidad_elementos_no_en_df_items = len(elementos_no_en_df_items)
    lista_elementos_no_en_df_items = list(elementos_no_en_df_items)

    print("Cantidad de items de la columna relacionada que no se encuentran en la columna principal ", (cantidad_elementos_no_en_df_items))
    print("Elementos que faltan en la columna principal:", lista_elementos_no_en_df_items)
    return lista_elementos_no_en_df_items

----

Las tablas:
- df_genres
- df_item_genre
- df_items
- df_reviews 
- df_users_items
- df_users 

estan relacionadas por lo que no deberíamos tener:
- df_items
    - items en df_item_genre que no esten en df_items
    - items en df_users_items que no esten en df_items
    - items en df_reviews que no esten en  df_items
- df_users
    - users en reviews que no esten en users
    - users en users_items que no esten en users

    - items en df_item_genre que no esten en df_items


In [5]:
items_de_df_item_genre_que_no_estan_en_df_items = items_faltantes(df_items.item_id,df_item_genre.item_id)

30754
28851
Cantidad de items de la columna relacionada que no se encuentran en la columna principal  1230
Elementos que faltan en la columna principal: [696350, 204840, 393270, 344140, 573520, 344150, 630870, 401500, 598110, 335970, 450660, 721000, 598120, 335981, 360560, 98421, 426110, 319620, 688274, 303250, 303251, 712870, 295100, 377020, 327870, 262341, 418000, 237780, 704740, 336100, 737510, 622820, 368870, 491760, 401650, 737530, 762120, 516360, 393500, 295210, 254260, 418110, 205120, 340, 319830, 434520, 426340, 754020, 369010, 311670, 328060, 737660, 278910, 598410, 672140, 467340, 344470, 459160, 393630, 410030, 901553, 328120, 344510, 696770, 770510, 352720, 377300, 369130, 369131, 672240, 737780, 737790, 393730, 713220, 98821, 303630, 606740, 320030, 729640, 508460, 664120, 426560, 246340, 295500, 688730, 303710, 754270, 295530, 688750, 361070, 598640, 443000, 746110, 402050, 574090, 680590, 311950, 426650, 459460, 377540, 525010, 525019, 508650, 353020, 271110, 385820, 328

    - items en df_users_items que no esten en df_items


In [6]:
items_de_df_users_items_que_no_estan_en_df_items = items_faltantes(df_items.item_id,df_users_items.item_id)

30754
10050
Cantidad de items de la columna relacionada que no se encuentran en la columna principal  1677
Elementos que faltan en la columna principal: [40960, 245760, 262150, 475150, 221200, 352280, 24600, 360480, 352290, 221220, 204840, 24620, 254000, 24630, 57400, 335930, 344130, 254020, 335950, 24660, 254040, 100, 41070, 368750, 344180, 221300, 417910, 213120, 516230, 393360, 458900, 368790, 221340, 8350, 360610, 8360, 385200, 221360, 8370, 254130, 24760, 8380, 352450, 295110, 98510, 418000, 286930, 336100, 24820, 221430, 49400, 41210, 524540, 24830, 368900, 319750, 24850, 24860, 377120, 360760, 16700, 418110, 442700, 57680, 340, 319830, 434520, 16730, 385370, 352610, 295270, 418150, 409960, 409970, 221560, 459130, 409980, 328060, 205180, 278910, 49540, 409990, 311700, 459160, 65950, 90530, 434600, 360880, 467380, 8640, 205250, 33220, 328140, 352720, 508380, 33250, 33260, 319980, 270830, 377330, 246260, 360950, 33270, 238070, 107000, 319990, 33280, 320000, 270850, 467460, 246280, 

    - items en df_reviews que no esten en  df_items


In [7]:
items_de_df_reviews_que_no_estan_en_df_items = items_faltantes(df_items.item_id,df_reviews.item_id)

30754
3488
Cantidad de items de la columna relacionada que no se encuentran en la columna principal  551
Elementos que faltan en la columna principal: [245760, 315392, 262150, 475150, 290830, 274450, 247830, 313372, 307230, 360480, 221220, 241700, 333860, 501800, 452650, 96300, 210990, 41010, 266290, 57400, 297020, 34880, 403520, 251970, 315460, 366660, 55370, 344140, 24660, 274520, 39000, 34910, 450660, 43110, 383080, 211050, 41070, 258160, 221300, 282740, 417910, 331900, 213120, 233610, 444560, 96400, 293010, 43160, 313500, 223390, 233630, 243870, 264360, 266410, 452780, 221360, 385200, 254130, 200900, 295110, 422100, 280790, 334040, 418010, 438490, 243930, 399580, 329950, 336100, 368870, 235760, 401650, 221430, 231670, 49400, 315640, 366844, 366845, 223490, 346370, 368900, 420110, 381210, 24860, 446750, 356640, 276770, 35110, 397610, 254260, 397620, 360760, 260410, 321860, 57680, 340, 92500, 317780, 270680, 373080, 16730, 385370, 323930, 256350, 358750, 227680, 295270, 430440, 20840

crearemos una lista de estos items

In [8]:
items_sobrantes = set(items_de_df_item_genre_que_no_estan_en_df_items + items_de_df_users_items_que_no_estan_en_df_items +items_de_df_reviews_que_no_estan_en_df_items)

In [9]:
len(items_sobrantes)

2867

In [10]:
df_items[df_items.item_id==245760]

Empty DataFrame
Columns: [item_id, title, url, release_date, developer, price]
Index: []

Existen 2867 items que se encuentran en las tablas relacionadas pero no en la tabla donde se almacena la info de items.
Para que la información sea consistente eliminaremos estos items de los df.

Eliminamos items de df_item_genre

In [11]:
df_item_genre[df_item_genre.item_id.isin(items_sobrantes)]

item_id      genres
32      768800      Casual
33      768800       Indie
34      768800      Racing
35      768800  Simulation
36      768570      Action
...        ...         ...
71528   771810      Action
71529   771810   Adventure
71530   771810       Indie
71535   769330      Casual
71536   769330       Indie

[3199 rows x 2 columns]

In [12]:
df_item_genre = df_item_genre[~(df_item_genre.item_id.isin(items_sobrantes))]

Eliminamos items de df_users_item

In [13]:
df_users_items[df_users_items.item_id.isin(items_sobrantes)]

user_id  item_id  playtime_forever
12       76561197970982479      340                37
16       76561197970982479     9340               692
31       76561197970982479    23120                53
34       76561197970982479    35010               570
40       76561197970982479    24860               437
...                    ...      ...               ...
3246314  76561198312638244   202990                21
3246315  76561198312638244   212910               490
3246337  76561198320038728    39000                71
3246342  76561198320136420   496920                 2
3246351  76561198329548331   519140                 3

[449797 rows x 3 columns]

In [14]:
df_users_items = df_users_items[~(df_users_items.item_id.isin(items_sobrantes))]

Eliminamos items de df_reviews

In [15]:
df_reviews[df_reviews.item_id.isin(items_sobrantes)]

user_id  item_id  recommend posted_date  year  \
2      76561197970982479    43110       True  2011-04-21  2011   
3                js41637   251610       True  2014-06-24  2014   
23     76561198089393905    72850       True  2015-02-01  2015   
42     76561198043472122    33440      False  2014-12-19  2014   
67     76561198054540475    12210       True  2015-02-13  2015   
...                  ...      ...        ...         ...   ...   
52169  76561198222628548   359800       True  2015-04-10  2015   
52188            wayfeng   369200       True         NaN     0   
52209  76561198267374962   369200       True  2015-12-29  2015   
52210  76561198270958927   381210       True         NaN     0   
52225  76561198296402247   280790       True         NaN     0   

       sentiment_analysis  
2                       2  
3                       2  
23                      0  
42                      0  
67                      2  
...                   ...  
52169                   0  
52188                   2  
52209                   2  
52210                   0  
52225                   0  

[5262 rows x 6 columns]

In [16]:
df_reviews = df_reviews[~(df_reviews.item_id.isin(items_sobrantes))]

    - users en reviews que no esten en users


In [17]:
items_de_df_reviews_que_no_estan_en_df_users = items_faltantes(df_users.user_id,df_reviews.user_id)

70912
21584
Cantidad de items de la columna relacionada que no se encuentran en la columna principal  0
Elementos que faltan en la columna principal: []


    - users en users_items que no esten en users

In [18]:
items_de_df_users_items_que_no_estan_en_df_users = items_faltantes(df_users.user_id,df_users_items.user_id)

70912
67829
Cantidad de items de la columna relacionada que no se encuentran en la columna principal  0
Elementos que faltan en la columna principal: []


Podemos ver que no existen usuarios en df_reviews y df_users_items que no se encuentren en df_users

Las tablas:
- df_genres
- df_item_genre
- df_items
- df_reviews 
- df_users_items
- df_users 

In [19]:
df_genres.to_csv("data/df_genres.csv",index=False)
df_item_genre.to_csv("data/df_item_genre.csv",index=False)
df_users_items.to_csv("data/df_users_items.csv",index=False)
df_reviews.to_csv("data/df_reviews.csv",index=False)
df_users_items.to_csv("data/df_users_items.csv",index=False)